### ISSUE TITLE
Chunked stream responses

### ISSUE THREAD

**Issue by alfsb** [justcoding121/Titanium-Web-Proxy#823]
Hi, I'm trying to read a long chunked response which never ends. So it's got blocked forever, with the proxy waiting for all chunks to assembly in one response body.

There is a way to inspect the chunked parts while delivering the responses in parallel?

If not, I suggest implementing a "Stream" flag or a "StreamSize" int, somewhere in `SessionEventArgs`, to make it possible to disable this caching mechanism, so that partial responses got delivered instantly.

A never ending response also calls for a "NoCache" flag, too, to avoid a OutOfMemoryException in long running connections.

**Comment by justcoding121** [comment]
If you don't call the response body ( for example  `await e.GetResponseBody() `) it should relay the chunked stream instead of waiting to read it to memory. 

@honfika In future, we may want to add an option to read the response body bytes by bytes by adding a new method. The problem is that in that case the user will have to write the response on his own, because proxy won't be caching the body. So user would have to call a new write to stream method. Also, once read from stream is called, user will not be allowed to call `SetBodyBytes()`. When write to stream is called for the first time, we need to fetch the connection to server and write the headers. Then, we write the bytes provided by user back to back. Something like below.

```
int readBytes = int.Max;
while(readBytes > 0)
{
 var bytesRead = await e.ReadResponseBodyStream(int bytesToRead, bool isChunked);
 readBytes = bytesRead.Length;

// User processes the bytes read on his own. (May be user writes the bytes to disc as a file)
 ... 

 await e.WriteResponseBodyStream(byte[] bytes, bool isChunked)
}

```
If someone could create a PR would be great.

**Comment by alfsb** [comment]
I think there are separate questions here.

1. A way to *inspect* all bytes received on the body, chunked or otherwise. Say, a drop copy interface, receives a callback, and that callback parameters contains all headers and the body or chunk bytes.

2. A way to change the received bytes. This could be accomplished on the same callback interface, where the bytes parameter, if changed or not, will be the bytes pushed on ResponseBody.

The problem above occurs only on chunked responses, but a clearer way to express these concepts is to have both interfaces, at same time:

a) a HttpMessageReceived callback, that contains all parsed headers and all bytes that are cached anyway.

b) a HttpChunkReceived callback, that contains only the bytes received on the chunk, with no cache by the proxy.

Both callbacks can have the same parameters, only the name denotes where this is a complete message or a simple chunk.

**Comment by justcoding121** [comment]
Based on the discussion above, I just added two new event handlers as a prototype to develop branch, which if someone could implement in future would be great.

https://github.com/justcoding121/titanium-web-proxy/commit/8a32fa99116afa82fad55476eef3ce161efad355

There are two scenarios that we need to handle.

1. **Read Only**: User do not modify the body bytes (chunked or otherwise). In that case, user can subscribe to the event handler shown below and peek the body bytes. The handlers will be called once for regular response with fixed content length header. It will be called multiple times for chunked response, i.e. each time a chunk is read and is about to be written. `IsFinalChunk` will be true when the chunk read from client or server stream is the last chunk.

2. **Read & Modify:** User wants to modify the body bytes. In that case, user would set the `byte[] BodyBytes { get; set; }` property. If `IsChunked ` is false, user will have to set the full body bytes only once.  If `IsChunked` is true, user would set chunk body and mark `IsFinalChunk` to true when it is the final chunk. In all cases, the handlers will be called again and again until `IsFinalChunk` is true. We would need to ensure that any unread chunked bytes are discarded from client or server stream by siphoning them out before finalizing the request or response. Also, if body is a compressed stream, that need to be handled properly, we need to decompress for the handler call, and compress upon write chunk call. 

Note that` IsChunked` property is readonly, i.e user will not be allowed to modify the` Transfer-encoding` or any other headers at this stage, because its already send to client or server. 

Similarly, in case of fixed length body with `Content-Length` header already sent, user is allowed to change body bytes, but its length should match exactly with `Content-Length` header already send. For fixed length body, user can use existing `BeforeRequest` or `BeforeResponse` event handlers to modify the headers first with different `Content-Length`, and then set the body bytes inside these handlers with its length matching exactly to the content-length header already sent.

![image](https://user-images.githubusercontent.com/2637679/115794678-a7724e80-a38b-11eb-8c7f-7a3a8f912854.png)

````CSharp
public class BeforeBodyWriteEventArgs : ProxyEventArgsBase
    {
        internal BeforeBodyWriteEventArgs(SessionEventArgs session, byte[] bodyBytes, bool isChunked, bool isFinalChunk) : base(session.ClientConnection)
        {
            Session = session;
            BodyBytes = bodyBytes;
            IsChunked = isChunked;
            IsFinalChunk = isFinalChunk;
        }


        /// <value>
        ///     The session arguments.
        /// </value>
        public SessionEventArgs Session { get; }

        /// <summary>
        ///  Indicates whether body is written chunked stream.
        ///  If this is true, BeforeRequestBodySend or BeforeResponseBodySend will be called until IsLastChunk is true.
        /// </summary>
        public bool IsChunked { get; }

        /// <summary>
        /// Indicates if this is the last chunk from client or server stream, when request is chunked.
        /// Override this property to true if there are more bytes to write.
        /// </summary>
        public bool IsFinalChunk { get; set; }

        /// <summary>
        /// The bytes about to be written. If IsChunked is true, this will be a chunk of the bytes to be written.
        /// Override this property with custom bytes if needed, and adjust IsLastChunk accordingly.
        /// </summary>
        public byte[] BodyBytes { get; set; }
    }
````

**Comment by justcoding121** [comment]
I think this approach is relatively simple to implement and  shouldn't break existing functionality. All we need to do is pass the new handlers all the way down to the write bytes call in http stream, and do the call back logic described above.

**Comment by justcoding121** [comment]
Actually, the above suggested new handlers can also be used for fixed content length body to be read bytes by bytes. In that case,  the handler will be called again and again each time we fill our read buffer, until content-length is reached. User can also set the bytes and the handler will be called again and again until content-length amount of bytes is reached. 

That would be helpful to keep the proxy memory footprint low, when the fixed content-length is large, say in tens of megabytes.

**Comment by justcoding121** [comment]
I've also did some prep work so that both `ProxyServer` and `SessionEventArgs` would be available inside `HttpStream`, so that everything needed for the event handlers will be in `HttpStream`.

**Comment by alfsb** [comment]
I'm suggesting to change the protocol a little, to make these new handlers
more akin to a filter than a batch processing. First I will comment inline,
and then the racionalle below.

On Thu, Apr 22, 2021 at 11:16 PM Jehonathan Thomas ***@***.***>
wrote:

>
>    1.
>
>    *Read & Modify:* User wants to modify the body bytes. In that case,
>    user would set the byte[] BodyBytes { get; set; } property.
>
> In that case, the user would set *or clear* the byte[] BodyBytes { get;
set; } property. (With null? With new byte[]? A method .ClearBytes() to
make this unambiguous?)

A cleared chunked fragment is not sent down or upstream if an intermediary
empty chunk message is impossible, or sent without any data if possible by
protocol.


>    1.
>
>    If IsChunked is false, user will have to set the full body bytes only
>    once.
>
> If IsChunked is false, the user will have a chance to modify body bytes
only once.


>
>    1.
>
>    If IsChunked is true, user would set chunk body and mark IsFinalChunk
>    to true when it is the final chunk.
>
> The user probably dont want to mess with IsFinalChunck, as this may cause
severe disruptions. Say, an intermediary chunk marked as final followed by
another chunk.


>
>    1.
>
>    In all cases, the handlers will be called again and again until
>    IsFinalChunk is false. We would need to ensure that any unread chunked
>    bytes are discarded from client or server stream by siphoning them out
>    before finalizing the request or response. Also, if body is a compressed
>    stream, that need to be handled properly, we need to decompress for the
>    handler call, and compress upon write chunk call.
>
> It's guaranteed that every chunk sits in a compression boundary or frame?
If not, this may not be possible without a ton of bookkeeping, up to the
original situation where a chunked message is eternally cached until the
RAM or a timeout runs off.

In this situation, a RawBytes/DecodedBytes property may be necessary in
lieu of the BodyBytes property that can be impossible to offer.


>    1.
>
>    Note that IsChunked property is readonly, i.e user will not be allowed
>    to modify the Transfer-encoding header at this stage, because its
>    already send to client or server.
>
> By the same reasoning, IsFinalChunk shouldn't be touched if the user
really *really* knows that he is doing, but I can foresee some
situations where the user need to mark some outbound chunk stream as
completed but keep collecting the inbound chunk stream.

So the protocol I suggest is this:

User wants to modify the body bytes. In that case, the user would set the
byte[] BodyBytes { get; set; } property in each callback, or set it to null
/ call .ClearBody() to avoid these bytes to be sent. IsChunked only
indicates if the original HTTP message is chunked or not, but this callback
is called on every body byte block received by the proxy, chunked or not.
The handlers will be called again and again until IsFinalChunk is *true*.
Also, if the body is a compressed stream, the RawBytes will contain the
original bytes, and ~BodyBytes~ DecodedBytes is filled with inflated data
on a best efforts basis, but may be null in case of errors. IsChunked
property is readonly, i.e user is not allowed to modify the
Transfer-encoding header at this stage, because it's already sent to client
or server.

André L F S Bacci

**Comment by alfsb** [comment]
On Thu, Apr 22, 2021 at 11:51 PM Jehonathan Thomas ***@***.***>
wrote:

> Actually, the above suggested new handlers can also be used for fixed
> content length body to be read bytes by bytes. In that case, the handler
> will be called again and again each time we fill our read buffer, until
> content-length is reached. User can also set the bytes and the handler will
> be called again and again until content-length amount of bytes is reached.
>
> That would be helpful to keep the proxy memory footprint low, when the
> fixed content-length is large, say in tens of megabytes.
>
Yes, there are some situations where this would be *necessary*. Say, a
proxy capable of recompressing or decompressing any request that passes
there, to fix some upstream or downstream incompatibility. Some big files
appear in this situation and not all RAM in the world will solve this
problem if it's necessary to wait for the whole file. And timeouts.

**Comment by justcoding121** [comment]
To clarify some of the things you mentioned.

**Sending Empty Body**
1. For chunked body, if you want to don't send any response body, you would set the `Bodybytes` to `null` or empty array and set `IsFinalChunk` to true. 
2. For Fixed body length,  if you want to don't send any response body, you would set the `Content-Length: 0` inside `OnBeforeRequest` or `OnBeforeResponse` handlers, and then set `BodyBytes` to null or empty array.

**Sending non-empty Body**
1. For chunked body, http compression is done for the entire stream. So when this handler is called, it will be uncompressed `BodyBytes` read from the underlying decompression stream. When writing the body back to server or client, it will then go through a compression stream. The type of compression stream (gzip, zlib etc) is based on  `content-encoding` header.
2. For fixed content-length, if you modify the body bytes, you need to specify the `content-length` of compressed modified body ahead in `OnBeforeRequest` or `OnBeforeResponse` handlers. This new handler will be called again and again until `content-length` number of bytes are send. Calculating compressed body length ahead may be difficult. To avoid that, one can mark this request or response as uncompressed inside `OnBeforeRequest` or `OnBeforeResponse` (use `content-encoding: identity`), then set `Content-Length` to the actual uncompressed byte length.

**Comment by justcoding121** [comment]
To clarify some of the things you mentioned.

**Sending Empty Body**
1. For chunked body, if you don't want to send any response body, you would set the `Content-Length: 0` and `transfer encoding: identity` inside `OnBeforeRequest` or `OnBeforeResponse` handlers. This new handler will still be called with chunks read (inside `BodyBytes`) from the client again and again, until all chunks are read. User will set `BodyBytes` to null or empty always.
2. For Fixed content length,  if you don't want to send any response body, you would set the `Content-Length: 0` inside `OnBeforeRequest` or `OnBeforeResponse` handlers. This handler will still be called with bytes read (inside `BodyBytes`) from the client again and again, until all original content-length bytes are read. User will set `BodyBytes` to null or empty always.

**Sending non-empty Body**
1. For chunked body, http compression is done for the entire stream. So when this handler is called, it will be uncompressed `BodyBytes` read from the underlying decompression stream. When writing the body back to server or client, it will then go through a compression stream automatically. The type of compression stream (gzip, zlib etc) is based on  `content-encoding` header, users don't need to worry about compression or decompression process.
2. For fixed content length, if you modify the body bytes, you need to specify the `content-length` of compressed modified body ahead in `OnBeforeRequest` or `OnBeforeResponse` handlers. This new handler will be called again and again until compressed `content-length` number of `BodyBytes` are send. Calculating compressed body length ahead may be difficult. To avoid that, one can mark this request or response as uncompressed inside `OnBeforeRequest` or `OnBeforeResponse` (use `content-encoding: identity`), then set `Content-Length` to the actual uncompressed byte length and provide uncompressed `BodyBytes` in this new handler.

**Comment by justcoding121** [comment]
To clarify some of the things you mentioned.

This new handlers will NOT be called if the body was already read through `GetRequestBodyAsString()` or `GetRequestBody()` from within existing before request handlers. In that case, user will have to call `SetRequestBody` or `SetRequestBodyString`. Same is true for response (i.e  `GetResponseBodyAsString()` or `GetResponseBody()`)

**Sending Empty Body**

**1. For chunked body**, if you don't want to send any response body, this new handler will be called with original chunks read (inside `BodyBytes`) from the client or server again and again, until all chunks are read. User will set `BodyBytes` to null or empty in return.

````CSharp
private async Task OnResponseBodyWrite(object sender, BeforeBodyWriteEventArgs e)
{
    //append original bytes to disk file 
    appendAndSaveOriginalBytes(e.BodyBytes);

    e.BodyBytes = null;
}
````

**2. For Fixed content length**,  if you don't want to send any response body, you would set the `Content-Length: 0` inside `OnBeforeRequest` or `OnBeforeResponse` handlers. This handler will still be called with original bytes read (inside `BodyBytes`) from the client or server again and again, until all original content-length bytes are read. User will return `BodyBytes` as null or empty always.

````CSharp
private async Task OnResponseBodyWrite(object sender, BeforeBodyWriteEventArgs e)
{
    //append original bytes to disk file 
    appendAndSaveOriginalBytes(e.BodyBytes);

    e.BodyBytes = null;
}
````

**Sending non-empty Body**

**1. For chunked body**, http compression is done for the entire stream. So when this handler is called, it will be uncompressed `BodyBytes` read from the underlying decompression stream of client or server. When writing the body back to server or client, it will then go through a compression stream automatically. The type of compression stream (gzip, zlib etc) is based on  `content-encoding` header, users don't need to worry about compression or decompression process. The new handler will be called again and again until `IsLastChunk` is true. 
  -  **1.a When modified chunked body is larger than original body:** The new handler will provide original body until its exhausted. After that it will send null as `BodyBytes`.
````CSharp
private async Task OnResponseBodyWrite(object sender, BeforeBodyWriteEventArgs e)
{
    //append original bytes to disk file 
   if(e.BodyBytes!=null){    appendAndSaveOriginalBytes(e.BodyBytes); }

   //set uncompressed new body bytes
    e.BodyBytes = modifiedBytes;
    
    if(modifiedBytes is LastChunk) { e.IsLastChunk = true; } else { e.IsLastChunk = false; }
}
````
  -  **1.b When modified chunked body is smaller than original body:** The new handler will set BodyBytes to null after its exhausted. 
````CSharp
private async Task OnResponseBodyWrite(object sender, BeforeBodyWriteEventArgs e)
{
    //append original bytes to disk file 
    appendAndSaveOriginalBytes(e.BodyBytes);

    if(all modified bytes are send) { e.BodyBytes = null; }    //set to null
    else { e.BodyBytes = modifiedBytes; }
}
````

**2. For fixed content length**, if you modify the body bytes, you need to specify the `content-length` of compressed modified body ahead in `OnBeforeRequest` or `OnBeforeResponse` handlers. This new handler will be called again and again at the minimum until original `content-length` number of `BodyBytes` are read. If the new content-length is larger than original, then this new handler will be called again and again with empty `BodyBytes` until new `content-length` is send. When the new content-length is smaller than the original content-length, user would set the `BodyBytes` to null for remaining callbacks. Calculating compressed body length ahead may be difficult. To avoid that, one can mark this request or response as uncompressed inside `OnBeforeRequest` or `OnBeforeResponse` (use `content-encoding: identity`), then set `Content-Length` to the actual uncompressed byte length and provide uncompressed `BodyBytes` in this new handler.
  -  **2.a When modified fixed length body is larger than original body:** The new handler will provide original body until its exhausted. After that it will send null as `BodyBytes`.
````CSharp
private async Task OnResponseBodyWrite(object sender, BeforeBodyWriteEventArgs e)
{
    //append original bytes to disk file 
   if(e.BodyBytes!=null){    appendAndSaveOriginalBytes(e.BodyBytes); }

   //set uncompressed new body bytes
    e.BodyBytes = modifiedBytes;
}
````
  -  **2.b When modified fixed length body is smaller than original body:** The new handler will set BodyBytes to null after its exhausted. 
````CSharp
private async Task OnResponseBodyWrite(object sender, BeforeBodyWriteEventArgs e)
{
    //append original bytes to disk file 
    appendAndSaveOriginalBytes(e.BodyBytes);

    //set to null if nothing more to send
    if(all modified bytes are send) { e.BodyBytes = null; } else { e.BodyBytes = modifiedBytes; }
}
````

**Comment by justcoding121** [comment]
Sorry for deleting my responses multiple times. 😅 I think I was thinking out loud while typing. 

One more thing that we need to do apart from this new handler is for `e.Ok()` calls inside "before request" or "before response" calls. In that case, we would still need to give an option for users to read the body as stream before canceling the request. That is out of scope for this work.

**Comment by zeltrax00** [comment]
Did anyone implement this feature ? Can I get a sample ?

**Comment by justcoding121** [comment]
This is implemented on current `develop` (5.6.0+) without the proposed `Stream` / `StreamSize` / `NoCache` flags. Bodies are relayed as they arrive by default; the proxy only buffers when you call `GetRequestBody()` / `GetResponseBody()`.

### Inspect chunks while delivering (do **not** call `GetResponseBody()`)

```csharp
proxyServer.OnResponseBodyWrite += (sender, e) =>
{
    // e.BodyBytes is this chunk / DATA frame (still Content-Encoded if the origin used gzip/br).
    // Assign a new array to change what the client receives.
    // e.IsLastChunk == true on the final piece.
    Console.WriteLine($"chunk {e.BodyBytes.Length} bytes, last={e.IsLastChunk}");
    return Task.CompletedTask;
};
```

Same idea for uploads: `OnRequestBodyWrite`.

Supported on HTTP/1.x, HTTP/2, and native HTTP/3 (per DATA frame). See [Streaming Bodies](https://github.com/justcoding121/titanium-web-proxy/wiki/Streaming-Bodies).

### Avoid OOM / hang on endless streams

- Prefer the hooks above — memory stays bounded by buffer/frame size.
- Whole-body reads are capped by `ProxyServer.MaxBufferedBodyBytes` (default 4 MiB).
- For a never-ending origin body you are short-circuiting, close instead of draining: `e.TerminateServerConnection()` or `RespondStreaming(..., closeServerConnection: true)`.

### Produce a synthetic streamed / endless body

```csharp
proxyServer.BeforeRequest += (sender, e) =>
{
    var response = new Response
    {
        StatusCode = 200,
        StatusDescription = "OK",
        HttpVersion = e.HttpClient.Request.HttpVersion
    };
    response.Headers.AddHeader(KnownHeaders.ContentType, "text/event-stream");

    e.RespondStreaming(response, async (stream, ct) =>
    {
        while (!ct.IsCancellationRequested)
        {
            var line = Encoding.UTF8.GetBytes($"data: {DateTime.UtcNow:o}\n\n");
            await stream.WriteAsync(line, ct);
            await Task.Delay(1000, ct);
        }
    }, closeServerConnection: true);

    return Task.CompletedTask;
};
```

Omit `Content-Length` for chunked (HTTP/1.1) or DATA-frame (HTTP/2/3) framing.

### Optional: decode / re-encode in the hook

Hooks see **wire bytes**. To send decompressed (or recompressed) content you must:

1. In `BeforeResponse`, strip/adjust `Content-Encoding` and drop `Content-Length` (use chunked / no length) **before** the body starts.
2. In `OnResponseBodyWrite`, keep a streaming decoder/encoder across chunks and replace `e.BodyBytes`; flush on `IsLastChunk`.

You cannot keep the original compressed `Content-Length` if the payload size changes.

**Comment by justcoding121** [comment]
### Follow-up: decompress / compress while streaming

`OnResponseBodyWrite` always sees **wire bytes** (still gzip/br if the origin set `Content-Encoding`). The proxy does not decompress for you.

**BCL note:** `GZipStream` / `DeflateStream` / `BrotliStream` in decompress mode are **read-based** — you cannot `Write` compressed bytes into them (that throws). A correct approach is either:

- **Simple (below):** buffer wire chunks, then on `IsLastChunk` decompress → optional edit → compress once and emit (fine for bounded responses; not for endless streams), or
- **True constant-memory streaming:** use a push-capable inflater (e.g. `BrotliDecoder.TryDecompress`, or a third-party zlib) across hook invocations.

```csharp
using System.IO.Compression;
using System.Text;
using Titanium.Web.Proxy.Models;

proxyServer.BeforeResponse += (_, e) =>
{
    var ce = e.HttpClient.Response.ContentEncoding;
    if (ce == null || !ce.Contains("gzip", StringComparison.OrdinalIgnoreCase))
        return Task.CompletedTask;

    // Recompressed size will not match the origin Content-Length.
    e.HttpClient.Response.Headers.RemoveHeader(KnownHeaders.ContentLength);
    if (e.HttpClient.Request.HttpVersion.Major < 2)
        e.HttpClient.Response.IsChunked = true; // H2/H3 use DATA frames; do not set TE: chunked

    e.UserData = new MemoryStream(); // accumulate gzip wire bytes
    return Task.CompletedTask;
};

proxyServer.OnResponseBodyWrite += (_, e) =>
{
    if (e.Session.UserData is not MemoryStream wire)
        return Task.CompletedTask;

    if (e.BodyBytes is { Length: > 0 })
        wire.Write(e.BodyBytes, 0, e.BodyBytes.Length);

    if (!e.IsLastChunk)
    {
        // Hold output until we can finish the gzip stream (BCL pull-based decompress).
        e.BodyBytes = Array.Empty<byte>();
        return Task.CompletedTask;
    }

    wire.Position = 0;
    byte[] plain;
    using (var gzip = new GZipStream(wire, CompressionMode.Decompress, leaveOpen: true))
    using (var plainMs = new MemoryStream())
    {
        gzip.CopyTo(plainMs);
        plain = plainMs.ToArray();
    }

    // Optional edit of decompressed content:
    // var text = Encoding.UTF8.GetString(plain);
    // plain = Encoding.UTF8.GetBytes(text.Replace("http://", "https://"));

    using (var outMs = new MemoryStream())
    {
        using (var gzip = new GZipStream(outMs, CompressionLevel.Fastest, leaveOpen: true))
            gzip.Write(plain, 0, plain.Length);
        e.BodyBytes = outMs.ToArray(); // still Content-Encoding: gzip
    }

    wire.Dispose();
    e.Session.UserData = null;
    return Task.CompletedTask;
};
```

**Send decompressed (no re-compress):** in `BeforeResponse` also remove `Content-Encoding`, and on `IsLastChunk` set `e.BodyBytes = plain` (skip the compress `GZipStream`).

**Do not** call `GetResponseBody()` in the same session — that buffers the body itself and skips the streaming hook. Same idea for requests via `BeforeRequest` + `OnRequestBodyWrite`. For `br` / `deflate`, swap in `BrotliStream` / `DeflateStream` the same way.
